systemd: disable NSCD when DNSSEC validation is disabled in timesyncd

nesteroff opened the pull request:
###### Description of changes

When a system has a significantly wrong date and time timesyncd is unable to synchronize it because DNSSEC doesn't work. In order to break this chicken and egg problem systemd-timesync disables DNSSEC validation by setting SYSTEMD_NSS_RESOLVE_VALIDATE=0 in the unit file. However, it doesn't work in NixOS because it uses NSCD.

This patch disables NSCD in systemd-timesyncd when SYSTEMD_NSS_RESOLVE_VALIDATE is set to 0 so that it uses NSS libraries directly. In order for it to be able to find the libnss_resolve.so.2 library it also adds the systemd directory to the LD_LIBRARY_PATH.

Fixes #174173

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nesteroff committed:
systemd: disable NSCD when DNSSEC validation is disabled in timesyncd

When a system has a wrong date and time timesyncd is unable to synchronize it
because DNSSEC doesn't work. In order to break this chicken and egg problem
systemd-timesync disables DNSSEC validation by setting
SYSTEMD_NSS_RESOLVE_VALIDATE=0 in the unit file. However, it doesn't work in
NixOS because it uses NSCD. This patch disables NSCD in systemd-timesyncd when
SYSTEMD_NSS_RESOLVE_VALIDATE is set to 0 so that it uses NSS libraries
directly. In order for it to be able to find the libnss_resolve.so.2 library
this patch adds the systemd directory in the nix store to the LD_LIBRARY_PATH.

nesteroff commented:
There was some discussion about this issue with @arianvp [here](https://github.com/systemd/systemd/pull/28110)

flokli commented:
This needs to target staging, please look at contributing.md to avoid people getting mass -pinged while changing branches.

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -365,6 +366,16 @@ stdenv.mkDerivation (finalAttrs: {
+  ''
+  # Set LD_LIBRARY_PATH so that the timesyncd service can find the libnss_resolve.so.2 library
+  + ''
+    substituteInPlace units/systemd-timesyncd.service.in \
```
We can move this into the 0020-*.patch directly, it'll be less brittle.

nesteroff replied:
Here I'm setting LD_LIBRARY_PATH to $out/lib so that it can find libnss_resolve.so.2 in thy systemd directory which is always different like /nix/store/7d5lp02lnh4kcivmwkwx75g0a1z06znw-systemd/lib/. Is there a way to add it to the 0020-*.patch?

flokli replied:
Yeah, there's some variables that'll get replaced by meson.

nesteroff replied:
Ah, ok, updated.

flokli commented:
Thanks! I wonder if there's any way we can test this in a VM test, but time obviously is complicated : 🤔

RaitoBezarius commented:
> Thanks! I wonder if there's any way we can test this in a VM test, but time obviously is complicated : thinking

Just `LD_PRELOAD` a preeny time library in systemd-timesyncd?

nesteroff commented:
> Thanks! I wonder if there's any way we can test this in a VM test, but time obviously is complicated : thinking

Yeah, unfortunately I'm not familiar with VM test but here is my manual test procedure, just in case:
- Stop timesyncd: sudo systemctl stop systemd-timesyncd
- Delete the clock file: sudo rm /var/lib/systemd/timesync/clock
- Set a wrong date and time: sudo date 01010101
- Start timesyncd: sudo systemctl start systemd-timesyncd
- DNS doesn't work and timesyncd is unable to get the correct date and time but with this patch it all works fine because timesyncd uses libnss_resolve.so.2 directly and it gets SYSTEMD_NSS_RESOLVE_VALIDATE=0.
- This is only applicable when DNSSEC is enabled.

ElvishJerricco commented:
I think this definitely needs a nixos style test of some kind. But also, is this a nixos specific thing? If not it definitely needs to be upstreamed and a link to the systemd PR needs to be commented.

arianvp commented:
Systemd closed the upstream PR.  See the discussion that was linked by OP

arianvp commented:
> still not grokking why you are not just using userdbd+resolved on nixos (i.e. ship nss-resolve + nss-systemd wherever you include glibc), but whatever. doubling down on nscd in 2022/2023 is so backwards, folks. In particular because glibc falls back to local NSS when nscd doesn't react after a very shot time-out. Which makes it useless for your case.
>
> Anyway, apparently nixos will solve this. Closing here.

RaitoBezarius commented:
> > still not grokking why you are not just using userdbd+resolved on nixos (i.e. ship nss-resolve + nss-systemd wherever you include glibc), but whatever. doubling down on nscd in 2022/2023 is so backwards, folks. In particular because glibc falls back to local NSS when nscd doesn't react after a very shot time-out. Which makes it useless for your case.

Do we know what we lack to actually do this? (also, you mentioned BDFL, what are you referring to?)

arianvp commented:
> > > still not grokking why you are not just using userdbd+resolved on nixos (i.e. ship nss-resolve + nss-systemd wherever you include glibc), but whatever. doubling down on nscd in 2022/2023 is so backwards, folks. In particular because glibc falls back to local NSS when nscd doesn't react after a very shot time-out. Which makes it useless for your case.
> 
> Do we know what we lack to actually do this? (also, you mentioned BDFL, what are you referring to?)

https://github.com/NixOS/nixpkgs/pull/111194 would allow us to implement this 

With bdfl i meant Eelco. He wants to come up with a solution to keep old binaries working without nscd. The linked PR is such a solution

nesteroff commented:
> Thanks! I wonder if there's any way we can test this in a VM test, but time obviously is complicated : 🤔

Here is a test idea that came to my mind:
- Enabled systemd-resolved with DNSSEC.
- Setup a dummy domain name using a simple local DNS server with no DNSSEC.
- Enable systemd-timesyncd and use the dummy domain as an NTP server.
- Enable debug log level in systemd-timesyncd and read the output.
- Without this patch it fails to resolve the domain name but when the patch is applied it successfully resolves it and tries to connect.

I’ve updated the commit with the test and rebased.

flokli commented on pkgs/os-specific/linux/systemd/0020-timesyncd-disable-NSCD-when-DNSSEC-validation-is-dis.patch:
```diff
@@ -0,0 +1,59 @@
+ # correct time to work, but we likely won't acquire that without NTP. Let's
+ # break this chicken-and-egg cycle here.
+ Environment=SYSTEMD_NSS_RESOLVE_VALIDATE=0
++Environment=LD_LIBRARY_PATH={{ROOTLIBDIR}}
```
We only need to start timesyncd with `LD_LIBRARY_PATH` set like this if systemd-resolved itself is also enabled (`services.resolved.enable = true;`) too, right?

Otherwise nothing will honor the `SYSTEMD_NSS_RESOLVE_VALIDATE` env var that's set.

I think I'd prefer if we set this in the NixOS module (in `nixos/modules/system/boot/timesyncd.nix`), and only if resolved is enabled, rather than patching it in uncondtionally:

```
systemd.services.systemd-timesyncd.environment.LD_LIBRARY_PATH = if config.services.resolved.enable "${config.systemd.package.out}/lib" else "";
```

The other part of the patch should obviously stay as-is.

nesteroff replied:
Actually, it looks like with the nssModules path added to systemd-timesyncd, everything works fine even without the patch because when NSCD fails to resolve the hostname, it falls back to modules from nsswitch.conf. In this case, systemd-resolved resolves the hostname fine because when loaded directly from timesyncd, it gets SYSTEMD_NSS_RESOLVE_VALIDATE=0. Disabling NSCD in systemd-timesyncd helps to save a few seconds, but now I'm not sure if it's worth it.

flokli replied:
The current approach is what I suggested, no? Keep the SYSTEMD_NSS_RESOLVE_VALIDATE=0 part in the patch, but configure the LD_LIBRARY_PATH env var through the module system.

nesteroff replied:
Yes, but my point is that if we use "if config.services.resolved.enable" then when resolved is not enabled, systemd-timesyncd will not be able to use any other NSS modules because NSCD is disabled in the patch and NSS modules are not in LD_LIBRARY_PATH.

flokli replied:
But nss-resolve is the only NSS module honoring `SYSTEMD_NSS_RESOLVE_VALIDATE=0` env var anyways, no?

nesteroff replied:
That's correct. I'm just saying that when nss-resolve is disabled we probably still want systemd-timesyncd to keep using other NSS modules that are configured in the system.

flokli replied:
Oh yes indeed! So `config.system.nssModules.path` makes more sense.

Can you extend the comment in the module system to make that point more clear?

nesteroff replied:
Sure, I tried to explain in the comment.

flokli replied:
LGTM!

flokli reviewed (commented):
Looking good!

I got the test to succeed with the patch applied, and verified it times out without the patch applied.

I think the VM test could be made a bit easier to understand and some comments could be added.
We also can move the LD_LIBRARY path env to a NixOS config to only apply when it's needed.

I put some comments inline.

nesteroff commented on nixos/tests/systemd-timesyncd-nscd-dnssec.nix:
```diff
@@ -0,0 +1,40 @@
+import ./make-test-python.nix ({ pkgs, ... }:
```
Sure, I tried to explain why it is needed and how it works. Feel free to suggest any corrections.

flokli replied:
LGTM, except for the fact it still mentions getent (commented there). Thanks!

nesteroff commented on nixos/tests/systemd-timesyncd-nscd-dnssec.nix:
```diff
@@ -0,0 +1,40 @@
+        ip = eth1IP;
+        data = ''
+          .ntp:${eth1IP}
+          +.fake.ntp:127.0.0.1:300
```
Never knew this RFC existed. I also tried to use a ".test" DNS name from RFC 2606, but for some reason, it resolves fine even though DNSSEC validation fails so I kept fake.ntp.

flokli replied:
Maybe there's some special-casing for `.test` somewhere, but I don't think it matters much in this internet-isolated VM test.

nesteroff replied:
Yep, `.test` is indeed a special case in systemd: https://github.com/systemd/systemd/blob/main/src/resolve/resolved-dns-trust-anchor.c#L112C22-L112C22. According to the documentation, negative trust anchors define domains where DNSSEC validation shall be turned off. I don’t think it should resolve such domains when `DNSSEC=true` but it looks like a similar bug have already been reported: https://github.com/systemd/systemd/issues/25676#issuecomment-1634810897

nesteroff commented on nixos/tests/systemd-timesyncd-nscd-dnssec.nix:
```diff
@@ -0,0 +1,40 @@
+    machine.wait_for_unit("tinydns.service")
+    machine.wait_for_unit("systemd-timesyncd.service")
+    machine.wait_until_succeeds("journalctl -u systemd-timesyncd.service --grep='Resolved address 127.0.0.1:123 for fake.ntp'", timeout=60)
+  '';
```
I added getent because it's already there and for nslookup we need to add pkgs.bind. I can switch no nslookup if you prefer it.

flokli replied:
`getent` has its own custom logic, and doesn't (always) use nsncd. I thought about using nslookup to the stub resolved ip, but we can probably just use  `resolvectl query`, which should be in `$PATH` already.

nesteroff replied:
Ok, updated to resolvectl.

nesteroff commented on nixos/tests/systemd-timesyncd-nscd-dnssec.nix:
```diff
@@ -0,0 +1,40 @@
+  testScript = ''
+    machine.wait_for_unit("tinydns.service")
+    machine.wait_for_unit("systemd-timesyncd.service")
+    machine.wait_until_succeeds("journalctl -u systemd-timesyncd.service --grep='Resolved address 127.0.0.1:123 for fake.ntp'", timeout=60)
```
Done

flokli commented on nixos/tests/systemd-timesyncd-nscd-dnssec.nix:
```diff
@@ -0,0 +1,61 @@
+# - Sets up a DNS server (tinydns) listening on the eth1 ip addess, serving .ntp and fake.ntp records.
+# - Configures that DNS server as a resolver and enables DNSSEC in systemd-resolved settings.
+# - Configures systemd-timesyncd to use fake.ntp hostname as an NTP server.
+# - Executes getent command to make sure DNS lookup fails due to broken DNSSEC.
```
This still mentions getent. Maybe just "performs a regular DNS lookup, to ensure it fails due to broken DNSSEC"?

nesteroff replied:
Good catch. Fixed.

flokli commented:
I successfully ran the test (again), and ensured it (still) fails (as expected) if the patch and `LD_LIBRARY_PATH` override are not present.

I did notice it also succeeds eventually with *just* `LD_LIBRARY_PATH` being set in the module system (I assume that's the nscd fallback after some timeout logic kicking in), but I still think having the patch is cleaner.

We know it can't work over nscd protocol in these cases, so we should disable it explicitly, rather than relying on that fallback-after-timeout behaviour in glibc.
